File: LeetCode/Medium/Search a 2D Matrix/Search a 2D Matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        
        if not matrix or target is None:
            return False
        
        cols, rows = len(matrix[0]), len(matrix)
        left, right = 0, cols*rows-1
        
        while left<=right:
            index = (left+right)//2
            num = matrix[index//cols][index%cols]
            
            if num == target:
                return True
            elif num > target:
                right = index-1
            else:
                left = index + 1
                
        return False

File: LeetCode/Medium/Search a 2D Matrix/Search a 2D Matrix.go
func searchMatrix(matrix [][]int, target int) bool {
    
    var check_idx int
    for idx, row := range matrix{
        compare := row[len(matrix[0])-1]
        if compare >= target{
            check_idx = idx
            break
        }
    }
    for _, val := range matrix[check_idx]{
        if val == target{
            return true
        }
    }
    return false
}
